### DISCUSSION TITLE
Multiple Concurrent Peppol Senders

### DISCUSSION THREAD

**Discussion by simon-temple** [phax/phase4#110] [Q&A] [answered]
The tests in https://github.com/phax/phase4/tree/master/phase4-peppol-client/src/test/java/com/helger/phase4/peppol show example senders that start with:
```
public static void main (final String [] args)
  {
    // Provide context
    WebScopeManager.onGlobalBegin (MockServletContext.create ());
```
 
If I add the WebScopeManager call to my DocumentSender class and I have two threads instantiate and use it at the same time I get an exception:
```
java.lang.IllegalStateException: Another global scope with ID 'd35e000877a711edba4d0242ac1e0003' is already present. New global scope with ID 'd3acbc4977a711edba4d0242ac1e0003' is not set!
 at com.helger.scope.mgr.ScopeManager.lambda$setGlobalScope$0(ScopeManager.java:102)
 at com.helger.commons.concurrent.SimpleReadWriteLock.writeLocked(SimpleReadWriteLock.java:244)
 at com.helger.scope.mgr.ScopeManager.setGlobalScope(ScopeManager.java:97)
 at com.helger.web.scope.mgr.WebScopeManager.onGlobalBegin(WebScopeManager.java:148)
 at com.helger.web.scope.mgr.WebScopeManager.onGlobalBegin(WebScopeManager.java:140)
```

How should I arrange my classes and use of WebScopeManager to allow concurrent peppol sending?  I'm guessing I might need to bind a `scope` to a thread/runner?

**Comment by phax** [comment, marked as answer]
@simon-temple please make sure the `WebScopeManager.onGlobalBegin (MockServletContext.create ());` (and the respective shutdown call) is only calles once per Java application.

So if you run the sender in multiple threads, see for example this code snippet as an example: 
https://github.com/phax/phase4/blob/master/phase4-peppol-client/src/test/java/com/helger/phase4/peppol/receivers/MainSendInParallelHelger.java
It refers to `MainPhase4PeppolSenderHelger.send` for the main action.

hth

**Reply by simon-temple** [reply]
@phax Thank you for the prompt response to my question :-)
